alpa/pipeline_parallel/stage_construction.py

```diff
@@ -82,7 +82,130 @@ def get_last_dp_result():
 
 
-@maybe_numba_jit
+@maybe_numba_jit(fastmath=True, nogil=True, cache=True)
+def get_optimal_submeshes(best_s, f_argmin, num_devices, num_layers, submesh_n_devices):
+    current_s = best_s
+    current_layer = 0
+    current_devices = num_devices
+
+    res = []
+    while current_s > 0 and current_layer < num_layers and current_devices > 0:
+        next_start_layer, submesh_choice, autosharding_choice = (
+            f_argmin[current_s, current_layer, current_devices])
+        assert next_start_layer != -1 and current_devices != -1
+        res.append(((current_layer, next_start_layer), submesh_choice,
+                    autosharding_choice))
+        current_s -= 1
+        current_layer = next_start_layer
+        current_devices -= submesh_n_devices[submesh_choice]
+    assert current_s == 0 and current_layer == num_layers and current_devices == 0
+
+    return res
+
+
+@maybe_numba_jit(fastmath=True, nogil=True, cache=True)
+def dp_impl_2(
+    num_layers, num_devices, submesh_sizes, valid_idxs_and_costs, max_n_succ_stages
+):
+    f = np.full(
+        (num_layers + 1, num_layers + 1, num_devices + 1), np.inf, dtype=np.float32
+    )
+    f_stage_max = np.full(
+        (num_layers + 1, num_layers + 1, num_devices + 1), 0.0, dtype=np.float32
+    )
+    f_argmin = np.full(
+        (num_layers + 1, num_layers + 1, num_devices + 1, 3), -1, dtype=np.int32
+    )
+    f[0, num_layers, 0] = 0
+    for d in maybe_prange(1, num_devices + 1):
+        for l, i, submesh_id, n_config, stage_cost in valid_idxs_and_costs:
+            l, i, submesh_id, n_config = map(int, (l, i, submesh_id, n_config))
+            n_submesh_devices = submesh_sizes[submesh_id]
+            if n_submesh_devices <= d:
+                for s in maybe_prange(1, num_layers + 1):
+                    if s - 1 > max_n_succ_stages[l, i, submesh_id, n_config]:
+                        continue
+
+                    new_cost = f[s - 1, i + 1, d - n_submesh_devices] + stage_cost
+                    if new_cost < f[s, l, d]:
+                        f[s, l, d] = new_cost
+                        f_argmin[s, l, d] = (i + 1, submesh_id, n_config)
+                        f_stage_max[s, l, d] = max(
+                            f_stage_max[s - 1, i + 1, d - n_submesh_devices], stage_cost
+                        )
+
+    return f, f_stage_max, f_argmin
+
+
+def dp_2(num_devices, num_microbatches, submesh_choices, compute_cost, max_n_succ_stages,
+):
+    """Auto stage dynamic programming."""
+    timers("stage-construction-dp").start()
+
+    num_layers = len(compute_cost)
+    all_possible_stage_costs = np.sort(np.unique(compute_cost))
+    best_cost = np.inf
+    best_solution = None
+    last_max_stage_cost = 0.0
+    # FIXME(zhuohan): Set this gap as a tunable parameter in global config
+    gap = 1e-6
+    assert len(
+        all_possible_stage_costs), "no solution in auto stage construction."
+
+    submesh_sizes: list = MaybeNumbaList()
```

Comment:
Why don't simply use an `np.array` here?

Reply:
yea that works too (`List` didn't work because numba couldn't infer the element types or something like that).

Reply:
actually i take it back, making the replacement fails type inference too:

```
No implementation of function Function(<built-in function getitem>) found for signature:
 
 >>> getitem(array(float32, 3d, C), Tuple(int64, int64, float64))
 
There are 22 candidate implementations:
   - Of which 20 did not match due to:
   Overload of function 'getitem': File: <numerous>: Line N/A.
     With argument(s): '(array(float32, 3d, C), Tuple(int64, int64, float64))':
    No match.
   - Of which 2 did not match due to:
   Overload in function 'GetItemBuffer.generic': File: numba/core/typing/arraydecl.py: Line 166.
     With argument(s): '(array(float32, 3d, C), Tuple(int64, int64, float64))':
    Rejected as the implementation raised a specific error:
      NumbaTypeError: unsupported array index type float64 in Tuple(int64, int64, float64)
  raised from /home/mlevental/miniconda3/envs/alpa/lib/python3.9/site-packages/numba/core/typing/arraydecl.py:72

During: typing of intrinsic-call at /home/mlevental/dev_projects/alpa/alpa/pipeline_parallel/stage_construction.py (129)

File "../../alpa/pipeline_parallel/stage_construction.py", line 129:
def dp_impl_2(
    <source elided>

                    new_cost = f[s - 1, i + 1, d - n_submesh_devices] + stage_cost
                    ^
```